### ISSUE TITLE
support Reader from multiple partitions

### ISSUE THREAD

**Issue by rogpeppe** [segmentio/kafka-go#355]
Currently there is a one-to-one relationship of Reader to partition, but a broker can be home to many partitions (potentially thousands). Kafka provides the bulk API calls for a reason, and issuing poll requests simultaneously over many TCP connections leads to significant network inefficiencies and increased load on the Kafka brokers.

It would be great to let a Reader read from multiple partitions on the same broker.

FWIW this issue is the main reason we're not moving to this package from sarama.

**Comment by robfig** [comment]
I came here to post the same issue - with 10k-partition topics, ignoring the network connection overhead, the approach of having a separate goroutine and buffered channel for each one seems like a major inefficiency.

**Comment by xiaofeige** [comment]
I agree with @rogpeppe . The reader here can only consume on partition, but the performance is pool , I have to create multiple reader to consume multiple partition by myself, this is not a good idea.

**Comment by hsawhney09** [comment]
Would an alternative be to use a consumer group while creating a Reader?

**Comment by rogpeppe** [comment]
@hsawhney09  how would that work?

**Comment by alessandronori** [comment]
> 
> 
> @hsawhney09 how would that work?

A workaround that I use is filling the `GroupID` field in `ReaderConfig`. In this way the `Reader` will fetch from every partition of a topic.
By the way I ended up here because I was looking for a way to set `StartOffset` even if there is a committed offset when you use `ConsumerGroup`. So also for me it would be nice to have a Reader that reads from multiple partitions without using `ConsumerGroup`.

**Comment by dominicbarnes** [comment]
As mentioned before, using a consumer group is one way to get more automatic consumption of topics + partitions, especially once #600 is merged and released.

If consumer groups aren't an option, creating a `kafka.Reader` for each partition is also an option and ostensibly isn't difficult to do. I'll work on adding an example of this exact scenario so we can test that assumption.

You also mentioned networking across many readers could be inefficient. I think the next logical step is to add support for `kafka.Transport` to `kafka.Reader`, which should allow connection sharing to address that concern. We can consider the bulk API in the future if further optimization is needed.
